### ISSUE TITLE
MacMini

### ISSUE THREAD

**Issue by 414726193** [nagadomi/nunif#112]
MPS: Unsupported Border padding mode

**Comment by nagadomi** [comment]
I don't have Mac hardware and it costs about $30/day for EC2 Mac instance, so I basically didn't test on macOS.
I will check it once within a year.

Maybe the error can be fixed as following. Not tested.
```patch
diff --git a/iw3/models/row_flow_v2.py b/iw3/models/row_flow_v2.py
index 0628521..1811c11 100644
--- a/iw3/models/row_flow_v2.py
+++ b/iw3/models/row_flow_v2.py
@@ -2,6 +2,7 @@ import torch
 import torch.nn as nn
 import torch.nn.functional as F
 from nunif.models import I2IBaseModel, register_model
+from collections import OrderedDict
 
 
 @register_model
@@ -14,15 +15,19 @@ class RowFlowV2(I2IBaseModel):
             nn.Conv2d(3, 16, kernel_size=(1, 3), stride=1, padding=(0, 1), padding_mode="replicate"),
             nn.ReLU(inplace=True))
         self.non_overlap = nn.Conv2d(16, 1, kernel_size=1, stride=1, padding=0)
-        self.overlap_residual = nn.Sequential(
-            nn.Conv2d(16, 16, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(16, 32, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(32, 32, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(32, 1, kernel_size=3, stride=1, padding=1, padding_mode="replicate"),
-        )
+        self.overlap_residual = nn.Sequential(OrderedDict([
+            ("pad0", nn.ReplicationPad2d((0, 4))),
+            ("0", nn.Conv2d(16, 16, kernel_size=(1, 9), stride=1, padding=0)),
+            ("1", nn.ReLU(inplace=True)),
+            ("pad1", nn.ReplicationPad2d((0, 4))),
+            ("2", nn.Conv2d(16, 32, kernel_size=(1, 9), stride=1, padding=0)),
+            ("3", nn.ReLU(inplace=True)),
+            ("pad2", nn.ReplicationPad2d((0, 4))),
+            ("4", nn.Conv2d(32, 32, kernel_size=(1, 9), stride=1, padding=0)),
+            ("5", nn.ReLU(inplace=True)),
+            ("pad3", nn.ReplicationPad2d((1, 1))),
+            ("6", nn.Conv2d(32, 1, kernel_size=3, stride=1, padding=0)),
+        ]))
         self.register_buffer("delta_scale", torch.tensor(1.0 / 127.0))
         self.delta_output = False
 ```

**Comment by 414726193** [comment]
Thank you BUT error：Only 2D, 3D, 4D, 5D padding with non-constant padding are supported for now

**Comment by nagadomi** [comment]
Sorry, I misunderstood `nn.ReplicationPad2d` argument.
```patch
diff --git a/iw3/models/row_flow_v2.py b/iw3/models/row_flow_v2.py
index 0628521..9085141 100644
--- a/iw3/models/row_flow_v2.py
+++ b/iw3/models/row_flow_v2.py
@@ -2,6 +2,7 @@ import torch
 import torch.nn as nn
 import torch.nn.functional as F
 from nunif.models import I2IBaseModel, register_model
+from collections import OrderedDict
 
 
 @register_model
@@ -14,15 +15,19 @@ class RowFlowV2(I2IBaseModel):
             nn.Conv2d(3, 16, kernel_size=(1, 3), stride=1, padding=(0, 1), padding_mode="replicate"),
             nn.ReLU(inplace=True))
         self.non_overlap = nn.Conv2d(16, 1, kernel_size=1, stride=1, padding=0)
-        self.overlap_residual = nn.Sequential(
-            nn.Conv2d(16, 16, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(16, 32, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(32, 32, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(32, 1, kernel_size=3, stride=1, padding=1, padding_mode="replicate"),
-        )
+        self.overlap_residual = nn.Sequential(OrderedDict([
+            ("pad0", nn.ReplicationPad2d((4, 4, 0, 0))),
+            ("0", nn.Conv2d(16, 16, kernel_size=(1, 9), stride=1, padding=0)),
+            ("1", nn.ReLU(inplace=True)),
+            ("pad1", nn.ReplicationPad2d((4, 4, 0, 0))),
+            ("2", nn.Conv2d(16, 32, kernel_size=(1, 9), stride=1, padding=0)),
+            ("3", nn.ReLU(inplace=True)),
+            ("pad2", nn.ReplicationPad2d((4, 4, 0, 0))),
+            ("4", nn.Conv2d(32, 32, kernel_size=(1, 9), stride=1, padding=0)),
+            ("5", nn.ReLU(inplace=True)),
+            ("pad3", nn.ReplicationPad2d((1, 1, 1, 1))),
+            ("6", nn.Conv2d(32, 1, kernel_size=3, stride=1, padding=0)),
+        ]))
         self.register_buffer("delta_scale", torch.tensor(1.0 / 127.0))
         self.delta_output = False
```

**Comment by 414726193** [comment]
Thank you for your answer, but the error is still there: MPS: Unsupported Border padding mode

**Comment by nagadomi** [comment]
Then MPS backend still cannot reproduce the same results as CUDA backend.
Use CPU, or wait until PyTorch dev implements replication padding for MPS, or implement replication padding yourself.


If you replace `nn.ReplicationPad2d` with `nn.ZeroPad2d` it maybe works, but the output of the model is undefined because the configuration is not the same as when the model was trained.

**Comment by nagadomi** [comment]
I use replication padding elsewhere so I will try to implement it for MPS. (It may be slower, but better than it not working)
Wait a few days if you are willing to test it.
But I don't know if it will work in the end, because even if I solve this problem, there could still be other problems.

**Comment by nagadomi** [comment]
Here is a naive implementation.
```patch
diff --git a/iw3/models/row_flow_v2.py b/iw3/models/row_flow_v2.py
index 0628521..9d82654 100644
--- a/iw3/models/row_flow_v2.py
+++ b/iw3/models/row_flow_v2.py
@@ -2,6 +2,29 @@ import torch
 import torch.nn as nn
 import torch.nn.functional as F
 from nunif.models import I2IBaseModel, register_model
+from collections import OrderedDict
+
+
+class ReplicationPad2dNaive(nn.Module):
+    def __init__(self, padding):
+        assert isinstance(padding, (list, tuple)) and len(padding) == 4
+        self.left = padding[0]
+        self.right = padding[1]
+        self.top = padding[2]
+        self.bottom = padding[3]
+        super().__init__()
+
+    def forward(self, x):
+        assert x.ndim == 4
+        if self.left > 0:
+            x = torch.cat((*((x[:, :, :, :1],) * self.left), x), dim=3)
+        if self.right > 0:
+            x = torch.cat((x, *((x[:, :, :, -1:],) * self.right)), dim=3)
+        if self.top > 0:
+            x = torch.cat((*((x[:, :, :1, :],) * self.top), x), dim=2)
+        if self.bottom > 0:
+            x = torch.cat((x, *((x[:, :, -1:, :],) * self.bottom)), dim=2)
+        return x
 
 
 @register_model
@@ -14,15 +37,19 @@ class RowFlowV2(I2IBaseModel):
             nn.Conv2d(3, 16, kernel_size=(1, 3), stride=1, padding=(0, 1), padding_mode="replicate"),
             nn.ReLU(inplace=True))
         self.non_overlap = nn.Conv2d(16, 1, kernel_size=1, stride=1, padding=0)
-        self.overlap_residual = nn.Sequential(
-            nn.Conv2d(16, 16, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(16, 32, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(32, 32, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(32, 1, kernel_size=3, stride=1, padding=1, padding_mode="replicate"),
-        )
+        self.overlap_residual = nn.Sequential(OrderedDict([
+            ("pad0", ReplicationPad2dNaive((4, 4, 0, 0))),
+            ("0", nn.Conv2d(16, 16, kernel_size=(1, 9), stride=1, padding=0)),
+            ("1", nn.ReLU(inplace=True)),
+            ("pad1", ReplicationPad2dNaive((4, 4, 0, 0))),
+            ("2", nn.Conv2d(16, 32, kernel_size=(1, 9), stride=1, padding=0)),
+            ("3", nn.ReLU(inplace=True)),
+            ("pad2", ReplicationPad2dNaive((4, 4, 0, 0))),
+            ("4", nn.Conv2d(32, 32, kernel_size=(1, 9), stride=1, padding=0)),
+            ("5", nn.ReLU(inplace=True)),
+            ("pad3", ReplicationPad2dNaive((1, 1, 1, 1))),
+            ("6", nn.Conv2d(32, 1, kernel_size=3, stride=1, padding=0)),
+        ]))
         self.register_buffer("delta_scale", torch.tensor(1.0 / 127.0))
         self.delta_output = False

```

**Comment by nagadomi** [comment]
Sorry, I missed one place, so I fixed it.
```patch
diff --git a/iw3/models/row_flow_v2.py b/iw3/models/row_flow_v2.py
index 0628521..18f9e50 100644
--- a/iw3/models/row_flow_v2.py
+++ b/iw3/models/row_flow_v2.py
@@ -2,6 +2,29 @@ import torch
 import torch.nn as nn
 import torch.nn.functional as F
 from nunif.models import I2IBaseModel, register_model
+from collections import OrderedDict
+
+
+class ReplicationPad2dNaive(nn.Module):
+    def __init__(self, padding):
+        assert isinstance(padding, (list, tuple)) and len(padding) == 4
+        self.left = padding[0]
+        self.right = padding[1]
+        self.top = padding[2]
+        self.bottom = padding[3]
+        super().__init__()
+
+    def forward(self, x):
+        assert x.ndim == 4
+        if self.left > 0:
+            x = torch.cat((*((x[:, :, :, :1],) * self.left), x), dim=3)
+        if self.right > 0:
+            x = torch.cat((x, *((x[:, :, :, -1:],) * self.right)), dim=3)
+        if self.top > 0:
+            x = torch.cat((*((x[:, :, :1, :],) * self.top), x), dim=2)
+        if self.bottom > 0:
+            x = torch.cat((x, *((x[:, :, -1:, :],) * self.bottom)), dim=2)
+        return x
 
 
 @register_model
@@ -14,17 +37,22 @@ class RowFlowV2(I2IBaseModel):
             nn.Conv2d(3, 16, kernel_size=(1, 3), stride=1, padding=(0, 1), padding_mode="replicate"),
             nn.ReLU(inplace=True))
         self.non_overlap = nn.Conv2d(16, 1, kernel_size=1, stride=1, padding=0)
-        self.overlap_residual = nn.Sequential(
-            nn.Conv2d(16, 16, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(16, 32, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(32, 32, kernel_size=(1, 9), stride=1, padding=(0, 4), padding_mode="replicate"),
-            nn.ReLU(inplace=True),
-            nn.Conv2d(32, 1, kernel_size=3, stride=1, padding=1, padding_mode="replicate"),
-        )
+        self.overlap_residual = nn.Sequential(OrderedDict([
+            ("pad0", ReplicationPad2dNaive((4, 4, 0, 0))),
+            ("0", nn.Conv2d(16, 16, kernel_size=(1, 9), stride=1, padding=0)),
+            ("1", nn.ReLU(inplace=True)),
+            ("pad1", ReplicationPad2dNaive((4, 4, 0, 0))),
+            ("2", nn.Conv2d(16, 32, kernel_size=(1, 9), stride=1, padding=0)),
+            ("3", nn.ReLU(inplace=True)),
+            ("pad2", ReplicationPad2dNaive((4, 4, 0, 0))),
+            ("4", nn.Conv2d(32, 32, kernel_size=(1, 9), stride=1, padding=0)),
+            ("5", nn.ReLU(inplace=True)),
+            ("pad3", ReplicationPad2dNaive((1, 1, 1, 1))),
+            ("6", nn.Conv2d(32, 1, kernel_size=3, stride=1, padding=0)),
+        ]))
         self.register_buffer("delta_scale", torch.tensor(1.0 / 127.0))
         self.delta_output = False
+        self.pre_pad = ReplicationPad2dNaive((28,) * 4)
 
         for m in self.modules():
             if isinstance(m, (nn.Conv2d,)):
@@ -66,7 +94,7 @@ class RowFlowV2(I2IBaseModel):
 
     def _forward_delta_only(self, x):
         assert not self.training
-        x = F.pad(x, [28] * 4, mode="replicate")
+        x = self.pre_pad(x)
         delta = self._forward(x)[1]
         delta = torch.cat([delta, torch.zeros_like(delta)], dim=1)
         delta = F.pad(delta, [-28] * 4)

```

**Comment by 414726193** [comment]
I'm sorry for some accidents these days, thank you for your answers. After I modified according to your suggestions, the same problem still occurred: MPS: Unsupported Border padding mode

**Comment by 414726193** [comment]
![image](https://github.com/nagadomi/nunif/assets/59109112/448280fb-b7d2-476d-a342-00d74eb8d860)
Change these two and you're ready to run

**Comment by nagadomi** [comment]
Thank you for the info.
That part is more of image post-processing rather than inference with machine learning model, so there is no major problem with that change. (The output image will be different from the CUDA version, but there is no practical problem.)

---
This is just my perspective.
Current code is 
https://github.com/nagadomi/nunif/blob/e42e2c57ad1d4163982efb3073ad5c6c2d9f15ee/iw3/utils.py#L247

 `padding_mode="border"` means the same as ReplicationPadding, the error makes sense.
Also, it is not that surprising that `mode="bicubic"` is not supported.
However, I did not expect `padding_mode="reflection"` to be supported. Because reflection padding is a more complex operation than replication padding. It is curious why PyTorch MPS backend does not support replication padding.

**Comment by 414726193** [comment]
Strange indeed. Here is another question, do you have a way to solve the problem from the root, thank you very much
> 
'aten::_upsample_bilinear2d_aa.out' is not currently implemented for the MPS device. If you want this op to be added in priority during the prototype phase of this feature, please comment on https://github.com/pytorch/pytorch/issues/77764. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.

**Comment by nagadomi** [comment]
Which line of which file causes that error? 

If you follow the error message, for example,
`PYTORCH_ENABLE_MPS_FALLBACK=1 python -m iw3.cli ...`
command seems to work around it (but it says to use CPU instead of MPS).
I am not familiar with IDE, but it should have environment variable settings.

**Comment by 414726193** [comment]
I set os.environ['PYTORCH_ENABLE_MPS_FALLBACK'] = '1' in code
Unable to run.
However, it works fine to type: export PYTORCH_ENABLE_MPS_FALLBACK=1 in the terminal
How would you set this environment variable in your code

**Comment by nagadomi** [comment]
Maybe it should be setup before importing torch.
In the **first line** of `iw3/cli.py` or `iw3/gui.py`, try adding
```python
import os
os.environ['PYTORCH_ENABLE_MPS_FALLBACK'] = '1'
```

**Comment by 414726193** [comment]
Wow! Success! Thank you! Thank you! Thank you very much, because I am a beginner and I can't do anything.
I want to package it and run it on mobile terminal, such as ipad pro with M1 chip. Is there any way

**Comment by nagadomi** [comment]
I don't know anything about Apple Devices and it's not my project, so just do it yourself.
In fact, I've never owned an Apple product.

**Comment by nagadomi** [comment]
The changes in this issue have already merged into master.

**Comment by dnzeek** [comment]
@414726193 

> Wow! Success! Thank you! Thank you! Thank you very much, because I am a beginner and I can't do anything. I want to package it and run it on mobile terminal, such as ipad pro with M1 chip. Is there any way

I too am interested in using iw3 on macOS. Could you please explain how you did it?

Thank you.

**Comment by nagadomi** [comment]
>The changes in this issue have already merged into master.

See https://github.com/nagadomi/nunif/blob/master/INSTALL-macos.md
After installation, you can launch iw3-gui with the following command
```
python3 -m iw3.gui
```
Also MPS is mostly slow compared to CUDA. https://github.com/nagadomi/nunif/discussions/327
And some models may not work https://github.com/nagadomi/nunif/issues/315#issuecomment-2691506422. I do not know how well it is currently working.

**Comment by dnzeek** [comment]
@nagadomi 

Thank you. I installed and am running it now on my Mac M2 Ultra using MPS.
It is converting at about 0.3 fps vs 0.5 fps on my PC Windows 11 RTX4090.
I will try later with CPU.

**Comment by nagadomi** [comment]
0.5 fps on RTX4090 is very slow than my senses.
Probably due to the high resolution of the video, the video encoding/decoding is taking up most of the processing time. It is more related to CPU and IO performance than GPU.

**Comment by dnzeek** [comment]
@nagadomi 

I spoke too soon. It started at 0.3 fps now it is down to 0.1 fps.

**Comment by dnzeek** [comment]
I think its due to the high resolution. Also I need a high bitrate so I am using "grain".

**Comment by dnzeek** [comment]
@nagadomi 

I think the GUI needs to be updated for libx265. The CRF range does not go high enough. According the the FFmpeg web page below the default for H.265 (libx265) is 28.

[https://trac.ffmpeg.org/wiki/Encode/H.265](url)

**Comment by nagadomi** [comment]
@dnzeek 
CRF editable and can enter up to 51 since it was pointed out before. https://github.com/nagadomi/nunif/discussions/236
The combox range is still adjusted to h264.

**Comment by dnzeek** [comment]
@nagadomi 

Oh. Thank you.
